### DISCUSSION TITLE
StyledComponent is not working in next js

### DISCUSSION THREAD

**Discussion by Clzzi** [vercel/next.js#36942] [Help] [answered]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.4.0: Fri Mar 18 00:46:32 PDT 2022; root:xnu-8020.101.4~15/RELEASE_ARM64_T6000
    Binaries:
      Node: 17.7.2
      npm: 8.5.2
      Yarn: 3.2.0
      pnpm: N/A
    Relevant packages:
      next: 12.1.6
      react: 18.1.0
      react-dom: 18.1.0

### What browser are you using? (if relevant)

Chrome 101.0.4951.64

### How are you deploying your application? (if relevant)

next dev OR next build & next start

### Describe the Bug

I set .babelrc file to use styled-components library.

Then, run next dev or next build & start. But i received an HTML file from the server that the style did not work.
However, after receiving the HTML file and hydrating becomes normal, style-components operate normally and the style works.
I tried to set the Babelrc file as below, but both results were the same.

```
{
  "presets": ["next/babel"],
  "plugins": [
    [
      "babel-plugin-styled-components",
      {
        "fileName": true,
        "displayName": true,
        "pure": true,
        "ssr": true
      }
    ]
  ]
}
```

```
{
  "presets": ["next/babel"],
  "plugins": [
    [
      "styled-components",
      {
        "ssr": true
      }
    ]
  ]
}

```

### Expected Behavior

I wanted the style to work normally on the HTML file that I received from the server after setting up the bablerc file.

### To Reproduce

in ```next dev```

![image](https://user-images.githubusercontent.com/62810965/168571718-fa8016bd-47d5-4e03-88d9-71c666285b56.png)

in ```next build & next start```

![image](https://user-images.githubusercontent.com/62810965/168571413-fba60c12-6940-428e-9970-43004c65c981.png)

**Comment by timneutkens** [comment]
You likely forgot to set up _document as per the example: https://github.com/vercel/next.js/blob/canary/examples/with-styled-components/pages/_document.js#L5-L29

**Reply by Clzzi** [reply]
I wrote the _document.tsx file as follows.

```typescript
import Document, {
  Html,
  Head,
  Main,
  NextScript,
  DocumentContext,
} from 'next/document';
import { ServerStyleSheet } from 'styled-components';

class MyDocument extends Document {
  static async getInitialProps(ctx: DocumentContext) {
    const sheet = new ServerStyleSheet();
    const originalRenderPage = ctx.renderPage;
    try {
      ctx.renderPage = () =>
        originalRenderPage({
          enhanceApp: (App) => (props) =>
            sheet.collectStyles(<App {...props} />),
        });

      const initialProps = await Document.getInitialProps(ctx);
      return {
        ...initialProps,
        styles: (
          <>
            {initialProps.styles}
            {sheet.getStyleElement()}
          </>
        ),
      };
    } finally {
      sheet.seal();
    }
  }

  render() {
    return (
      <Html lang="en">
        <Head>
          <link
            rel="preload"
            as="style"
            href="https://fonts.googleapis.com/css2?family=Noto+Sans+KR&display=swap"
          />
          <link
            href="https://fonts.googleapis.com/css2?family=Noto+Sans+KR&display=swap"
            rel="stylesheet"
          />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}

export default MyDocument;
```

**Comment by kmanadkat** [comment, marked as answer]
With Next v12, few lines of addition in `next.config.js` fixes the problem.

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  compiler: {
    styledComponents: true,          // ADD this compiler object
  },
}

module.exports = nextConfig
```
